mlir/lib/Dialect/Rock/Transforms/LowerRockReduce.cpp

```diff
@@ -0,0 +1,233 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ============================================================
+//
+// This pass converts rock.reduce into the TransformingFor loop
+// using global atomics.
+//
+//===-----------------------------------------------------===//
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/Rock/IR/Rock.h"
+#include "mlir/Dialect/Rock/IR/TransformMapBuilder.h"
+#include "mlir/Dialect/Rock/Passes.h"
+#include "mlir/Dialect/Rock/utility/transformMapUtils.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Transforms/RegionUtils.h"
+
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "llvm/Support/Debug.h"
+#include <memory>
+
+namespace mlir {
+namespace rock {
+#define GEN_PASS_DEF_ROCKLOWERREDUCEPASS
+#include "mlir/Dialect/Rock/Passes.h.inc"
+} // namespace rock
+} // namespace mlir
+
+#define DEBUG_TYPE "rock-lower-rock-reduce"
+
+using namespace mlir;
+using namespace mlir::rock;
+
+namespace {
+class RockLowerReducePass
+    : public rock::impl::RockLowerReducePassBase<RockLowerReducePass> {
+  void runOnOperation() override;
+};
+
+struct ReduceRewritePattern : public OpConversionPattern<ReduceOp> {
+  using OpConversionPattern<ReduceOp>::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(ReduceOp op, ReduceOpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override;
+};
+} // end namespace
+
+// This function will create a view of the tensor that is being
+// reduced from arbitary dimension sizes to bid x tid x iter space.
+static Value createThreadView(Value redInput, int64_t blockSize,
+                              int64_t gridSize, Location loc,
+                              PatternRewriter &rewriter) {
+  auto totalThreads = gridSize * blockSize;
+  auto inpShape = redInput.getType().cast<ShapedType>();
+  auto elementCount = inpShape.getNumElements();
+  auto dataPerThread = (elementCount + (totalThreads - 1)) / totalThreads;
+
+  llvm::errs() << "totalThreads = " << totalThreads << "\n";
+  llvm::errs() << "elementCount = " << elementCount << "\n";
+  llvm::errs() << "dataPerThread = " << dataPerThread << "\n";
+
+  SmallVector<SmallString<8>> topNames;
+  SmallVector<int64_t> topSizes;
+  int64_t topSizeProduct = 1;
+  SmallVector<unsigned int> topDims;
+  SmallVector<StringRef> topNameRefs;
+  for (auto dimAndSize : llvm::enumerate(inpShape.getShape())) {
```

Comment:
There's already a three-argument constructor for these builders so you don't have to do this.

Reply:
Can you eloborate what you mean here ? Is there another way to do this or are you saying the builder of the op can capture this ?

Reply:
`TransformMapBuilder` has a constructor that takes three arguments with no names, just a shape.

And if you need the names, you can call `getLowerName(i)` or possibly `getLowerNames()` to get the names - I'd look at the `BottomUpTMBuilder` to find the methods.

Reply:
OK, well I had to migrate a protected one to be public; I think it should be okay ? (as we have getEndNames anyway).

Reply:
@sjw36: FYI, now we can use this API elsewhere we used to create names for TrMaps when the dims are dynamic.

Reply:
I don't think you want `getEndNames()`, you might want `getStartNames()`

Reply:
And the reason `getStartNames()` was protected was because I didn't want people trying to modify that list and break class invariants.

We could add `getStartNameRefs()` populating a `SmallVectorImpl<StringRef>&`

Reply:
Thats what I have added here ?

```
void TransformMapBuilder::getStartNames(SmallVectorImpl<StringRef> &names) {
```